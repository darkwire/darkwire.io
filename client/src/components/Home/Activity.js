import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Message from 'components/Message'
import Username from 'components/Username'
import Notice from 'components/Notice'
import Zoom from 'utils/ImageZoom'
import { getObjectUrl } from 'utils/file'

import T from 'components/T'

class Activity extends Component {
  constructor(props) {
    super(props)

    this.state = {
      zoomableImages: [],
    }
  }

  getFileDisplay(activity) {
    const type = activity.fileType
    if (type.match('image.*')) {
      return (
        <img
          ref={c => this._zoomableImage = c}
          className="image-transfer zoomable"
          src={`data:${activity.fileType};base64,${activity.encodedFile}`}
          alt={`${activity.fileName} from ${activity.username}`}
          onLoad={this.handleImageDisplay.bind(this)}
        />
      )
    }
    return null
  }

  handleImageDisplay() {
    Zoom(this._zoomableImage)
    this.props.scrollToBottom()
  }

  getActivityComponent(activity) {
    switch (activity.type) {
      case 'TEXT_MESSAGE':
        return (
          <Message
            sender={activity.username}
            message={activity.text}
            timestamp={activity.timestamp}
          />
        )
      case 'USER_ENTER':
        return (
          <Notice>
            <div>
              <T data={{
                username: <Username key={0} username={activity.username} />
              }} path='userJoined'/>
            </div>
          </Notice>
        )
      case 'USER_EXIT':
        return (
          <Notice>
            <div>
              <T data={{
                username: <Username key={0} username={activity.username} />
              }} path='userLeft'/>
            </div>
          </Notice>
        )
      case 'TOGGLE_LOCK_ROOM':
        if (activity.locked) {
          return (
            <Notice>
              <div><T data={{
                username: <Username key={0} username={activity.username} />
              }} path='lockedRoom'/></div>
            </Notice>
          )
        } else {
          return (
            <Notice>
              <div><T data={{
                username: <Username key={0} username={activity.username} />
              }} path='unlockedRoom'/></div>
            </Notice>
          )
        }
      case 'NOTICE':
        return (
          <Notice>
            <div>{activity.message}</div>
          </Notice>
        )
      case 'CHANGE_USERNAME':
        return (
          <Notice>
            <div><T data={{
              oldUsername: <Username key={0} username={activity.currentUsername} />,
              newUsername: <Username key={1} username={activity.newUsername} />
            }} path='nameChange'/>
            </div>
          </Notice>
        )
      case 'USER_ACTION':
        return (
          <Notice>
            <div>&#42; <Username username={activity.username} /> {activity.action}</div>
          </Notice>
        )
      case 'RECEIVE_FILE':
        const downloadUrl = getObjectUrl(activity.encodedFile, activity.fileType)
        return (
          <div>
            <T data={{
              username: <Username key={0} username={activity.username} />,
            }} path='userSentFile'/>&nbsp;

            <a target="_blank" href={downloadUrl} rel="noopener noreferrer" download={activity.fileName}>
              <T data={{
                filename: activity.fileName,
              }} path='downloadFile'/>
            </a>
            {this.getFileDisplay(activity)}
          </div>
        )
      case 'SEND_FILE':
        const url = getObjectUrl(activity.encodedFile, activity.fileType)
        return (
          <Notice>
            <div>
              <T data={{
                filename: <a key={0} target="_blank" href={url} rel="noopener noreferrer" download={activity.fileName}>{activity.fileName}</a>,
              }} path='sentFile'/>&nbsp;
            </div>
            {this.getFileDisplay(activity)}
          </Notice>
        )
      default:
        return false
    }
  }

  render() {
    return (
      this.getActivityComponent(this.props.activity)
    )
  }
}

Activity.propTypes = {
  activity: PropTypes.object.isRequired,
  scrollToBottom: PropTypes.func.isRequired,
}

export default Activity;
