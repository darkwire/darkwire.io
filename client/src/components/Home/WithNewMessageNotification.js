import React, { Component } from 'react';
import { connect } from 'react-redux';
import { notify, beep } from 'utils/notifications';
import Tinycon from 'tinycon';

const mapStateToProps = state => {
  return {
    activities: state.activities.items,
    unreadMessageCount: state.app.unreadMessageCount,
    windowIsFocused: state.app.windowIsFocused,
    soundIsEnabled: state.app.soundIsEnabled,
    notificationIsEnabled: state.app.notificationIsEnabled,
    room: state.room,
  };
};

const WithNewMessageNotification = WrappedComponent => {
  return connect(mapStateToProps)(
    class WithNotificationHOC extends Component {
      state = { lastMessage: null, unreadMessageCount: 0 };

      static getDerivedStateFromProps(nextProps, prevState) {
        const {
          room: { id: roomId },
          activities,
          notificationIsEnabled,
          soundIsEnabled,
          unreadMessageCount,
          windowIsFocused,
        } = nextProps;

        if (activities.length === 0) {
          return null;
        }

        const lastMessage = activities[activities.length - 1];
        const { username, text } = lastMessage;

        if (lastMessage !== prevState.lastMessage && !windowIsFocused) {
          const title = `Message from ${username} (${roomId})`;
          if (notificationIsEnabled) notify(title, text);
          if (soundIsEnabled) beep.play();
        }

        if (unreadMessageCount !== prevState.unreadMessageCount) {
          Tinycon.setBubble(unreadMessageCount);
        }

        return { lastMessage, unreadMessageCount };
      }

      render() {
        // Filter props
        const {
          room,
          activities,
          notificationIsEnabled,
          soundIsEnabled,
          unreadMessageCount,
          windowIsFocused,
          ...rest
        } = this.props;
        return <WrappedComponent {...rest} />;
      }
    },
  );
};

export default WithNewMessageNotification;
