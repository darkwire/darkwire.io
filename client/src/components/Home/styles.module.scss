.styles {
  .tos {
    text-align: center;
    a {
      color: #666;
    }
  }
}
