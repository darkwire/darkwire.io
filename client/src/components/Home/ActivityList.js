import React from 'react'
import PropTypes from 'prop-types'
import ChatInput from 'components/Chat'
import Activity from './Activity'
import T from 'components/T'
import { defer } from 'lodash'

import styles from './styles.module.scss'

const ActivityList = ({ activities, openModal }) => {
  const [focusChat, setFocusChat] = React.useState(false);
  const [scrolledToBottom, setScrolledToBottom] = React.useState(true);
  const messageStream = React.useRef(null);
  const activitiesList = React.useRef(null);

  React.useEffect(() => {
    const currentMessageStream = messageStream.current;

    // Update scrolledToBottom state if we scroll the activity stream
    const onScroll = () => {
      const messageStreamHeight = messageStream.current.clientHeight
      const activitiesListHeight = activitiesList.current.clientHeight

      const bodyRect = document.body.getBoundingClientRect()
      const elemRect = activitiesList.current.getBoundingClientRect()
      const offset = elemRect.top - bodyRect.top
      const activitiesListYPos = offset

      const newScrolledToBottom = (activitiesListHeight + (activitiesListYPos - 60)) <= messageStreamHeight
      if (newScrolledToBottom) {
        if (!scrolledToBottom) {
          setScrolledToBottom(true)
        }
      } else if (scrolledToBottom) {
        setScrolledToBottom(false)
      }
    }

    currentMessageStream.addEventListener('scroll', onScroll)
    return () => {
      // Unbind event if component unmounted
      currentMessageStream.removeEventListener('scroll', onScroll)
    }
  }, [scrolledToBottom])

  const scrollToBottomIfShould = React.useCallback(() => {
    if (scrolledToBottom) {
      messageStream.current.scrollTop = messageStream.current.scrollHeight
    }
  }, [scrolledToBottom])

  React.useEffect(() => {
    scrollToBottomIfShould(); // Only if activities.length bigger
  }, [scrollToBottomIfShould, activities]);

  const scrollToBottom = React.useCallback(() => {
    messageStream.current.scrollTop = messageStream.current.scrollHeight
    setScrolledToBottom(true)
  }, [])

  const handleChatClick = () => {
    setFocusChat(true);
    defer(() => setFocusChat(false))
  }

  return (
    <div className="main-chat">
      <div onClick={handleChatClick} className="message-stream h-100" ref={messageStream} data-testid="main-div">
        <ul className="plain" ref={activitiesList}>
          <li><p className={styles.tos}><button className='btn btn-link' onClick={() => openModal('About')}> <T path='agreement'/></button></p></li>
          {activities.map((activity, index) => (
            <li key={index} className={`activity-item ${activity.type}`}>
              <Activity activity={activity} scrollToBottom={scrollToBottomIfShould} />
            </li>
          ))}
        </ul>
      </div>
      <div className="chat-container">
        <ChatInput scrollToBottom={scrollToBottom} focusChat={focusChat} />
      </div>
    </div>
  )
}

ActivityList.propTypes = {
  activities: PropTypes.array.isRequired,
  openModal: PropTypes.func.isRequired,
}

export default ActivityList
