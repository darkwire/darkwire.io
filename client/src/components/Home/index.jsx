import { connect } from 'react-redux';
import { useLoaderData } from 'react-router-dom';

import {
  receiveEncryptedMessage,
  createUser,
  openModal,
  closeModal,
  toggleWindowFocus,
  toggleSoundEnabled,
  togglePersistenceEnabled,
  toggleNotificationEnabled,
  toggleNotificationAllowed,
  toggleSocketConnected,
  receiveUnencryptedMessage,
  sendUnencryptedMessage,
  sendEncryptedMessage,
  setLanguage,
} from '@/actions';

import Home from './Home';
import WithNewMessageNotification from './WithNewMessageNotification';

const mapStateToProps = state => {
  const me = state.room.members.find(m => m.id === state.user.id);

  return {
    activities: state.activities.items,
    userId: state.user.id,
    username: state.user.username,
    publicKey: state.user.publicKey,
    privateKey: state.user.privateKey,
    members: state.room.members.filter(m => m.username && m.publicKey),
    roomId: state.room.id,
    roomLocked: state.room.isLocked,
    modalComponent: state.app.modalComponent,
    iAmOwner: Boolean(me && me.isOwner),
    faviconCount: state.app.unreadMessageCount,
    soundIsEnabled: state.app.soundIsEnabled,
    persistenceIsEnabled: state.app.persistenceIsEnabled,
    notificationIsEnabled: state.app.notificationIsEnabled,
    notificationIsAllowed: state.app.notificationIsAllowed,
    socketConnected: state.app.socketConnected,
    language: state.app.language,
    translations: state.app.translations,
  };
};

const mapDispatchToProps = {
  receiveEncryptedMessage,
  createUser,
  openModal,
  closeModal,
  toggleWindowFocus,
  toggleSoundEnabled,
  togglePersistenceEnabled,
  toggleNotificationEnabled,
  toggleNotificationAllowed,
  toggleSocketConnected,
  receiveUnencryptedMessage,
  sendUnencryptedMessage,
  sendEncryptedMessage,
  setLanguage,
};

export const ConnectedHome = WithNewMessageNotification(connect(mapStateToProps, mapDispatchToProps)(Home));

const HomeWithParams = ({ ...props }) => {
  const socketId = useLoaderData();
  return <ConnectedHome socketId={socketId} {...props} />;
};

export default HomeWithParams;
