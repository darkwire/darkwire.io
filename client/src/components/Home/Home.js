import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Crypto from 'utils/crypto'
import { connect as connectSocket } from 'utils/socket'
import Nav from 'components/Nav'
import shortId from 'shortid'
import Connecting from 'components/Connecting'
import Modal from 'react-modal'
import About from 'components/About'
import Settings from 'components/Settings'
import Welcome from 'components/Welcome'
import RoomLocked from 'components/RoomLocked'
import { X, AlertCircle } from 'react-feather'
import Tinycon from 'tinycon'
import beepFile from 'audio/beep.mp3'
import classNames from 'classnames'
import ActivityList from './ActivityList'

import styles from './styles.module.scss'

const crypto = new Crypto()

Modal.setAppElement('#root');

class Home extends Component {

  async componentWillMount() {
    const roomId = encodeURI(this.props.match.params.roomId)

    const user = await this.createUser()

    const socket = connectSocket(roomId)

    this.socket = socket;

    socket.on('disconnect', () => {
      this.props.toggleSocketConnected(false)
    })

    socket.on('connect', () => {
      this.initApp(user)
      this.props.toggleSocketConnected(true)
    })

    socket.on('USER_ENTER', (payload) => {
      this.props.receiveUnencryptedMessage('USER_ENTER', payload)
      this.props.sendEncryptedMessage({
        type: 'ADD_USER',
        payload: {
          username: this.props.username,
          publicKey: this.props.publicKey,
          isOwner: this.props.iAmOwner,
          id: this.props.userId,
        },
      })
      if (payload.users.length === 1) {
        this.props.openModal('Welcome');
      }
    })

    socket.on('USER_EXIT', (payload) => {
      this.props.receiveUnencryptedMessage('USER_EXIT', payload)
    })

    socket.on('ENCRYPTED_MESSAGE', (payload) => {
      this.props.receiveEncryptedMessage(payload)
    })

    socket.on('TOGGLE_LOCK_ROOM', (payload) => {
      this.props.receiveUnencryptedMessage('TOGGLE_LOCK_ROOM', payload)
    })

    socket.on('ROOM_LOCKED', (payload) => {
      this.props.openModal('Room Locked')
    });

    window.addEventListener('beforeunload', (evt) => {
      socket.emit('USER_DISCONNECT')
    });
  }

  componentDidMount() {
    this.bindEvents()

    this.beep = window.Audio && new window.Audio(beepFile)
  }

  componentWillReceiveProps(nextProps) {
    Tinycon.setBubble(nextProps.faviconCount)

    if (nextProps.faviconCount !== 0 && nextProps.faviconCount !== this.props.faviconCount && this.props.soundIsEnabled) {
      this.beep.play()
    }
  }

  getModal() {
    switch (this.props.modalComponent) {
      case 'Connecting':
        return {
          component: <Connecting />,
          title: 'Connecting...',
          preventClose: true,
        }
      case 'About':
        return {
          component: <About roomId={this.props.roomId} />,
          title: this.props.translations.aboutHeader,
        }
      case 'Settings':
        return {
          component: <Settings roomId={this.props.roomId} toggleSoundEnabled={this.props.toggleSoundEnabled} soundIsEnabled={this.props.soundIsEnabled} setLanguage={this.props.setLanguage} language={this.props.language} translations={this.props.translations} />,
          title: this.props.translations.settingsHeader,
        }
      case 'Welcome':
        return {
          component: <Welcome roomId={this.props.roomId} close={this.props.closeModal} translations={this.props.translations} />,
          title: this.props.translations.welcomeHeader,
        }
      case 'Room Locked':
        return {
          component: <RoomLocked modalContent={this.props.translations.lockedRoomHeader} />,
          title: this.props.translations.lockedRoomHeader,
          preventClose: true,
        }
      default:
        return {
          component: null,
          title: null,
        }
    }
  }

  initApp(user) {
    this.socket.emit('USER_ENTER', {
      publicKey: user.publicKey,
    })
  }

  bindEvents() {
    window.onfocus = () => {
      this.props.toggleWindowFocus(true)
    }

    window.onblur = () => {
      this.props.toggleWindowFocus(false)
    }
  }

  createUser() {
    return new Promise(async (resolve) => {
      const username = shortId.generate()

      const encryptDecryptKeys = await crypto.createEncryptDecryptKeys()
      const exportedEncryptDecryptPrivateKey = await crypto.exportKey(encryptDecryptKeys.privateKey)
      const exportedEncryptDecryptPublicKey = await crypto.exportKey(encryptDecryptKeys.publicKey)

      this.props.createUser({
        username,
        publicKey: exportedEncryptDecryptPublicKey,
        privateKey: exportedEncryptDecryptPrivateKey,
      })

      resolve({
        publicKey: exportedEncryptDecryptPublicKey,
      })
    })
  }

  render() {
    const modalOpts = this.getModal()
    return (
      <div className={classNames(styles.styles, 'h-100')}>
        <div className="nav-container">
          {!this.props.socketConnected &&
            <div className="alert-banner">
              <span className="icon"><AlertCircle size="15" /></span> <span>Disconnected</span>
            </div>
          }
          <Nav
            members={this.props.members}
            roomId={this.props.roomId}
            roomLocked={this.props.roomLocked}
            toggleLockRoom={() => this.props.sendUnencryptedMessage('TOGGLE_LOCK_ROOM')}
            openModal={this.props.openModal}
            iAmOwner={this.props.iAmOwner}
            userId={this.props.userId}
            translations={this.props.translations}
          />
        </div>
        <ActivityList
          openModal={this.props.openModal}
          activities={this.props.activities}
        />
        <Modal
          isOpen={Boolean(this.props.modalComponent)}
          contentLabel="Modal"
          style={{ overlay: { zIndex: 10 } }}
          className={{
            base: 'react-modal-content',
            afterOpen: 'react-modal-content_after-open',
            beforeClose: 'react-modal-content_before-close',
          }}
          overlayClassName={{
            base: 'react-modal-overlay',
            afterOpen: 'react-modal-overlay_after-open',
            beforeClose: 'react-modal-overlay_before-close',
          }}
          shouldCloseOnOverlayClick={!modalOpts.preventClose}
          onRequestClose={this.props.closeModal}
        >
          <div className="react-modal-header">
            {!modalOpts.preventClose &&
              <button onClick={this.props.closeModal} className="btn btn-link btn-plain close-modal">
                <X />
              </button>
            }
            <h3 className="react-modal-title">
              {modalOpts.title}
            </h3>
          </div>
          <div className="react-modal-component">
            {modalOpts.component}
          </div>
        </Modal>
      </div>
    )
  }
}

Home.defaultProps = {
  modalComponent: null,
}

Home.propTypes = {
  receiveEncryptedMessage: PropTypes.func.isRequired,
  receiveUnencryptedMessage: PropTypes.func.isRequired,
  createUser: PropTypes.func.isRequired,
  activities: PropTypes.array.isRequired,
  username: PropTypes.string.isRequired,
  publicKey: PropTypes.object.isRequired,
  members: PropTypes.array.isRequired,
  match: PropTypes.object.isRequired,
  roomId: PropTypes.string.isRequired,
  roomLocked: PropTypes.bool.isRequired,
  modalComponent: PropTypes.string,
  openModal: PropTypes.func.isRequired,
  closeModal: PropTypes.func.isRequired,
  iAmOwner: PropTypes.bool.isRequired,
  userId: PropTypes.string.isRequired,
  toggleWindowFocus: PropTypes.func.isRequired,
  faviconCount: PropTypes.number.isRequired,
  soundIsEnabled: PropTypes.bool.isRequired,
  toggleSoundEnabled: PropTypes.func.isRequired,
  toggleSocketConnected: PropTypes.func.isRequired,
  socketConnected: PropTypes.bool.isRequired,
  sendUnencryptedMessage: PropTypes.func.isRequired,
  sendEncryptedMessage: PropTypes.func.isRequired
}

export default Home;
