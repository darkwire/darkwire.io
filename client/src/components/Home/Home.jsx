import React from 'react';
import ReactModal from 'react-modal';
import PropTypes from 'prop-types';
import { nanoid } from 'nanoid';
import { X, AlertCircle } from 'react-feather';
import classNames from 'classnames';
import { useSelector, useDispatch } from 'react-redux';
import { createUser } from '@/reducers/user';

import Crypto from '@/utils/crypto';
import { connect as connectSocket } from '@/utils/socket';

import Nav from '@/components/Nav';
import Connecting from '@/components/Connecting';
import About from '@/components/About';
import Settings from '@/components/Settings';
import Welcome from '@/components/Welcome';
import RoomLocked from '@/components/RoomLocked';

import ActivityList from './ActivityList';

import styles from './styles.module.scss';

const crypto = new Crypto();

ReactModal.setAppElement('#root');

const Modal = ({
  closeModal,
  modalComponent,
  roomId,
  translations,
  toggleSoundEnabled,
  togglePersistenceEnabled,
  soundIsEnabled,
  persistenceIsEnabled,
  toggleNotificationEnabled,
  toggleNotificationAllowed,
  notificationIsEnabled,
  notificationIsAllowed,
  setLanguage,
  language,
}) => {
  const getModal = () => {
    switch (modalComponent) {
      case 'Connecting':
        return {
          component: <Connecting />,
          title: 'Connecting...',
          preventClose: true,
        };
      case 'About':
        return {
          component: <About roomId={roomId} />,
          title: translations.aboutHeader,
        };
      case 'Settings':
        return {
          component: (
            <Settings
              roomId={roomId}
              toggleSoundEnabled={toggleSoundEnabled}
              togglePersistenceEnabled={togglePersistenceEnabled}
              soundIsEnabled={soundIsEnabled}
              persistenceIsEnabled={persistenceIsEnabled}
              toggleNotificationEnabled={toggleNotificationEnabled}
              toggleNotificationAllowed={toggleNotificationAllowed}
              notificationIsEnabled={notificationIsEnabled}
              notificationIsAllowed={notificationIsAllowed}
              setLanguage={setLanguage}
              language={language}
              translations={translations}
            />
          ),
          title: translations.settingsHeader,
        };
      case 'Welcome':
        return {
          component: <Welcome roomId={roomId} close={closeModal} translations={translations} />,
          title: translations.welcomeHeader,
        };
      case 'Room Locked':
        return {
          component: <RoomLocked modalContent={translations.lockedRoomHeader} />,
          title: translations.lockedRoomHeader,
          preventClose: true,
        };
      default:
        return {
          component: null,
          title: null,
        };
    }
  };

  const modalOpts = getModal();
  return (
    <ReactModal
      isOpen={Boolean(modalComponent)}
      contentLabel="Modal"
      style={{ overlay: { zIndex: 10 } }}
      className={{
        base: 'react-modal-content',
        afterOpen: 'react-modal-content_after-open',
        beforeClose: 'react-modal-content_before-close',
      }}
      overlayClassName={{
        base: 'react-modal-overlay',
        afterOpen: 'react-modal-overlay_after-open',
        beforeClose: 'react-modal-overlay_before-close',
      }}
      shouldCloseOnOverlayClick={!modalOpts.preventClose}
      onRequestClose={closeModal}
    >
      <div className="react-modal-header">
        {!modalOpts.preventClose && (
          <button onClick={closeModal} className="btn btn-link btn-plain close-modal">
            <X />
          </button>
        )}
        <h3 className="react-modal-title">{modalOpts.title}</h3>
      </div>
      <div className="react-modal-component">{modalOpts.component}</div>
    </ReactModal>
  );
};

const Home = ({
  receiveEncryptedMessage,
  receiveUnencryptedMessage,
  activities,
  username,
  publicKey,
  members,
  socketId,
  roomId,
  roomLocked,
  modalComponent,
  openModal,
  closeModal,
  iAmOwner,
  userId,
  toggleWindowFocus,
  soundIsEnabled,
  persistenceIsEnabled,
  toggleSoundEnabled,
  togglePersistenceEnabled,
  notificationIsEnabled,
  notificationIsAllowed,
  toggleNotificationEnabled,
  toggleNotificationAllowed,
  toggleSocketConnected,
  socketConnected,
  sendUnencryptedMessage,
  sendEncryptedMessage,
  translations,
  setLanguage,
  language,
}) => {
  const socketPayloadRef = React.useRef({
    username: username,
    publicKey: publicKey,
    isOwner: iAmOwner,
    id: userId,
  });
  socketPayloadRef.current = {
    username: username,
    publicKey: publicKey,
    isOwner: iAmOwner,
    id: userId,
  };

  // Add blur et focus listeners
  React.useEffect(() => {
    const onFocus = () => {
      toggleWindowFocus(true);
    };
    const onBlur = () => {
      toggleWindowFocus(false);
    };

    window.addEventListener('focus', onFocus);
    window.addEventListener('blur', onBlur);

    return () => {
      window.removeEventListener('focus', onFocus);
      window.removeEventListener('blur', onBlur);
    };
  }, [toggleWindowFocus]);

  React.useEffect(() => {
    const socket = connectSocket(socketId);

    socket.on('disconnect', () => {
      toggleSocketConnected(false);
    });

    socket.on('connect', () => {
      socket.emit('USER_ENTER', {
        publicKey: socketPayloadRef.current.publicKey,
      });
      toggleSocketConnected(true);
    });

    socket.on('USER_ENTER', payload => {
      receiveUnencryptedMessage('USER_ENTER', payload);
      sendEncryptedMessage({
        type: 'ADD_USER',
        payload: socketPayloadRef.current,
      });
      if (payload.users.length === 1) {
        openModal('Welcome');
      }
    });

    socket.on('USER_EXIT', payload => {
      receiveUnencryptedMessage('USER_EXIT', payload);
    });

    socket.on('ENCRYPTED_MESSAGE', payload => {
      receiveEncryptedMessage(payload);
    });

    socket.on('TOGGLE_LOCK_ROOM', payload => {
      receiveUnencryptedMessage('TOGGLE_LOCK_ROOM', payload);
    });

    socket.on('ROOM_LOCKED', () => {
      openModal('Room Locked');
    });

    const onUnload = () => {
      socket.emit('USER_DISCONNECT');
    };

    window.addEventListener('beforeunload', onUnload);

    return () => {
      window.removeEventListener('beforeunload', onUnload);
      onUnload();
      socket.close();
    };
  }, [
    openModal,
    receiveEncryptedMessage,
    receiveUnencryptedMessage,
    sendEncryptedMessage,
    socketId,
    toggleSocketConnected,
  ]);

  return (
    <div className={classNames(styles.styles, 'h-100')}>
      <div className="nav-container">
        {!socketConnected && (
          <div className="alert-banner">
            <span className="icon">
              <AlertCircle size="15" />
            </span>{' '}
            <span>Disconnected</span>
          </div>
        )}
        <Nav
          members={members}
          roomId={roomId}
          roomLocked={roomLocked}
          toggleLockRoom={() => sendUnencryptedMessage('TOGGLE_LOCK_ROOM')}
          openModal={openModal}
          iAmOwner={iAmOwner}
          userId={userId}
          translations={translations}
        />
      </div>
      <ActivityList openModal={openModal} activities={activities} />
      <Modal
        closeModal={closeModal}
        modalComponent={modalComponent}
        roomId={roomId}
        translations={translations}
        toggleSoundEnabled={toggleSoundEnabled}
        togglePersistenceEnabled={togglePersistenceEnabled}
        soundIsEnabled={soundIsEnabled}
        persistenceIsEnabled={persistenceIsEnabled}
        toggleNotificationEnabled={toggleNotificationEnabled}
        toggleNotificationAllowed={toggleNotificationAllowed}
        notificationIsEnabled={notificationIsEnabled}
        notificationIsAllowed={notificationIsAllowed}
        setLanguage={setLanguage}
        language={language}
      />
    </div>
  );
};

export const WithUser = ({ ...rest }) => {
  const [loaded, setLoaded] = React.useState(false);
  const loading = React.useRef(false);

  const user = useSelector(state => state.user);
  const dispatch = useDispatch();

  React.useEffect(() => {
    let mounted = true;

    const createUserLocal = async () => {
      const localUsername = user.username || nanoid();

      const encryptDecryptKeys = await crypto.createEncryptDecryptKeys();
      const exportedEncryptDecryptPrivateKey = await crypto.exportKey(encryptDecryptKeys.privateKey);
      const exportedEncryptDecryptPublicKey = await crypto.exportKey(encryptDecryptKeys.publicKey);

      if (!mounted) {
        loading.current = false;
        return;
      }

      const payload = {
        username: localUsername,
        publicKey: exportedEncryptDecryptPublicKey,
        privateKey: exportedEncryptDecryptPrivateKey,
      };
      dispatch(createUser(payload));

      dispatch({ type: 'CREATE_USER', payload });

      loading.current = false;
      setLoaded(true);
    };

    if (!loaded && !loading.current) {
      loading.current = true;
      createUserLocal();
    }

    return () => {
      loading.current = false;
      mounted = false;
    };
  }, [dispatch, loaded, user.username]);

  if (!loaded) {
    return null;
  }

  return <Home username={user.username} publicKey={user.publicKey} userId={user.id} {...rest} />;
};

WithUser.defaultProps = {
  modalComponent: null,
};

WithUser.propTypes = {
  receiveEncryptedMessage: PropTypes.func.isRequired,
  receiveUnencryptedMessage: PropTypes.func.isRequired,
  activities: PropTypes.array.isRequired,
  members: PropTypes.array.isRequired,
  socketId: PropTypes.string.isRequired,
  roomId: PropTypes.string.isRequired,
  roomLocked: PropTypes.bool.isRequired,
  modalComponent: PropTypes.string,
  openModal: PropTypes.func.isRequired,
  closeModal: PropTypes.func.isRequired,
  iAmOwner: PropTypes.bool.isRequired,
  toggleWindowFocus: PropTypes.func.isRequired,
  soundIsEnabled: PropTypes.bool.isRequired,
  persistenceIsEnabled: PropTypes.bool.isRequired,
  toggleSoundEnabled: PropTypes.func.isRequired,
  togglePersistenceEnabled: PropTypes.func.isRequired,
  notificationIsEnabled: PropTypes.bool.isRequired,
  notificationIsAllowed: PropTypes.bool,
  toggleNotificationEnabled: PropTypes.func.isRequired,
  toggleNotificationAllowed: PropTypes.func.isRequired,
  toggleSocketConnected: PropTypes.func.isRequired,
  socketConnected: PropTypes.bool.isRequired,
  sendUnencryptedMessage: PropTypes.func.isRequired,
  sendEncryptedMessage: PropTypes.func.isRequired,
  setLanguage: PropTypes.func.isRequired,
  language: PropTypes.string.isRequired,
  translations: PropTypes.object.isRequired,
};

export default WithUser;
