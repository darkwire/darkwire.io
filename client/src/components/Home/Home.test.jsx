import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { test, expect, vi } from 'vitest';

import Home from './Home';
import configureStore from '@/store';

const store = configureStore();

vi.mock('react-modal'); // Cant load modal without root app element

vi.mock('@/RoomLink');
vi.mock('@/components/Nav');

vi.mock('@/utils/socket', () => {
  // Avoid exception
  return {
    connect: vi.fn().mockImplementation(() => {
      return {
        on: vi.fn(),
        emit: vi.fn(),
      };
    }),
  };
}); //

vi.mock('../../utils/crypto', () => {
  // Need window.crytpo.subtle
  return {
    default: vi.fn().mockImplementation(() => {
      return {
        createEncryptDecryptKeys: () => {
          return {
            privateKey: 'private',
            publicKey: 'public',
          };
        },
        exportKey: () => {
          return 'exportedkey';
        },
      };
    }),
  };
});

test('Home component is displaying', async () => {
  const { asFragment } = render(
    <Provider store={store}>
      <Home
        translations={{}}
        members={[]}
        openModal={() => {}}
        activities={[]}
        match={{ params: { roomId: 'roomTest' } }}
        createUser={() => {}}
        toggleSocketConnected={() => {}}
        receiveEncryptedMessage={() => {}}
        receiveUnencryptedMessage={() => {}}
        scrolledToBottom={true}
        setScrolledToBottom={() => {}}
        iAmOwner={true}
        roomLocked={false}
        userId={'userId'}
        roomId={'testId'}
        sendEncryptedMessage={() => {}}
        sendUnencryptedMessage={() => {}}
        socketConnected={false}
        toggleSoundEnabled={() => {}}
        soundIsEnabled={false}
        toggleNotificationEnabled={() => {}}
        toggleNotificationAllowed={() => {}}
        notificationIsEnabled={false}
        faviconCount={0}
        toggleWindowFocus={() => {}}
        closeModal={() => {}}
        publicKey={{}}
        username={'linus'}
      />
    </Provider>,
  );

  expect(asFragment()).toMatchSnapshot();
});
