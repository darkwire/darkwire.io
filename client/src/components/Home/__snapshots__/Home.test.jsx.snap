// Vitest Snapshot v1

exports[`Home component is displaying 1`] = `
<DocumentFragment>
  <div
    class="_styles_0b54d3 h-100"
  >
    <div
      class="nav-container"
    >
      <div
        class="alert-banner"
      >
        <span
          class="icon"
        >
          <svg
            fill="none"
            height="15"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
            <line
              x1="12"
              x2="12"
              y1="8"
              y2="12"
            />
            <line
              x1="12"
              x2="12.01"
              y1="16"
              y2="16"
            />
          </svg>
        </span>
         
        <span>
          Disconnected
        </span>
      </div>
    </div>
    <div
      class="main-chat"
    >
      <div
        class="message-stream h-100"
        data-testid="main-div"
      >
        <ul
          class="plain"
        >
          <li>
            <p
              class="_tos_0b54d3"
            >
              <button
                class="btn btn-link"
              >
                 By using Darkwire, you are agreeing to our Acceptable Use Policy and Terms of Service
              </button>
            </p>
          </li>
        </ul>
      </div>
      <div
        class="chat-container"
      >
        <form
          class="chat-preflight-container"
        >
          <textarea
            class="chat"
            placeholder="Type here"
            rows="1"
          />
          <div
            class="input-controls"
          >
            <div
              class="_styles_374fdd icon file-transfer btn btn-link"
            >
              <input
                id="fileInput"
                name="fileUploader"
                placeholder="Choose a file..."
                type="file"
              />
              <label
                for="fileInput"
              >
                <svg
                  fill="none"
                  height="24"
                  stroke="#fff"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M13 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V9z"
                  />
                  <polyline
                    points="13 2 13 9 20 9"
                  />
                </svg>
              </label>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
